src/lang/base/PrimTypes.ml

```diff
@@ -126,10 +126,21 @@ let string_of_uint_lit = function
   | Uint128L i' -> Uint128.to_string i'
   | Uint256L i' -> Uint256.to_string i'
 
+(* Scilla only allows printable ASCII characters from 32 (0x20) up-to 126 (0x7e).
+ * The only exceptions allowed are various whitespaces: \b \f \n \r \t. *)
+let validate_string_literal s =
+  let specials = ['\b'; '\012'; '\n'; '\r'; '\t'] in
+  String.fold s ~init:true ~f:(fun safe c ->
+    if not safe then false else
+    let c_int = Char.to_int c in
+    if c_int >= 32 && c_int <= 126 then true else
```

Comment:
I'm probably missing something here, but can this check not be performed in the lexer instead?

Reply:
It can be, but we also need the check for strings coming in from JSONs (particularly, message JSONs). JSON to literal conversion functions call `build_prim_literal`, thus also validating the constructed string literals.

So it is necessary here anyway, and since also sufficient, I didn't modify the lexer for these checks.